torch/csrc/nnpack/NNPACK.cpp

```diff
@@ -0,0 +1,348 @@
+#include "NNPACK.h"
+
+#include "TH/TH.h"
+#include <stdlib.h>
+
+namespace torch {
+namespace nnpack {
+
+// Stolen from Caffe2
+static pthreadpool_t nnpack_threadpool_ = nullptr;
+
+pthreadpool_t nnpack_threadpool() {
+  if (nnpack_threadpool_ == nullptr) {
+    enum nnp_status nnpack_status = nnp_initialize();
+    if (nnpack_status != nnp_status_success) throw std::runtime_error("could not initialize NNPack");
+    nnpack_threadpool_ = pthreadpool_create(16);
+  }
+  return nnpack_threadpool_;
+}
+
+// Assuming for now this is Thread-safe, but that seems like a stretch
+static void *workspace = nullptr;
+static size_t workspace_size = 0;
+
+// NNPack has alignment requirements
+const size_t nnpack_memory_alignment_boundary = 64;
+
+static inline void deallocate_workspace() {
+  if (workspace)
+    std::free(workspace);
+  workspace = nullptr;
+}
+
+static inline void allocate_workspace() {
+  if (workspace)
+    deallocate_workspace();
+  posix_memalign(&workspace, nnpack_memory_alignment_boundary, workspace_size);
```

Comment:
use `std::aligned_alloc`

Reply:
I think `std::aligned_alloc` is C++17? http://en.cppreference.com/w/cpp/memory/c/aligned_alloc

Reply:
Right, my bad. `posix_memalign` will unfortunately break Windows, so it would be nicer to use some standard function. Perhaps we could allocate `workspace_size + nnpack_memory_alignment_boundary` bytes and align them ourselves?

Reply:
I don't think NNPACK supports Windows either, so I don't think this will be a problem for now?

Reply:
Can we explicitly disable NNPack on Windows then, to avoid compilation errors on Windows users? cc @peterjc123